js/web/lib/wasm/jsep/webgpu/types.ts

```diff
@@ -26,64 +26,96 @@ export interface TensorInfo {
 
 
 export interface ProgramVariable {
-  type: 'float'|'int';
-  name: string;
-  arrayLength?: number;
-  data: number|number[];
+  type: 'int32'|'float32'|'uint32';
+  data: number|readonly number[];
 }
 
+/**
+ * Represent the dependency of a program on a specific input tensor.
+ *
+ * - 'none': the shader/uniform does not depend on this input's info
+ * - 'type': the shader/uniform depends on data type of this input
+ * - 'rank': the shader/uniform depends on data type and the rank of this input
+ * - 'dims': the shader/uniform depends on data type and the dims of this input
+ * - 'data': the shader/uniform depends on data type, the dims and the data of this input
+ */
+export type ProgramInputTensorInfoDependency = 'none'|'type'|'rank'|'dims'|'data';
 
-export interface ProgramMetadata {
+/**
+ * Represent information about a program's cache for shader.
+ */
+export interface ProgramShaderCacheInfo {
   /**
-   * the name of the program. used for debugging and profiling
+   * an optional string as a cache hint in the artifact cache. If this is not specified, the cache hint will be empty.
+   *
+   * This hint string should only contains initializing-time information, such as the attributes or any information of
+   * initializers. It should NOT contain any runtime information, such as the shape of inputs.
    */
-  name: string;
+  hint?: string;
 
   /**
-   * gpu data types for each input
-   */
-  inputTypes: GpuDataType[];
-  /**
-   * an optional string as a cache hint in the artifact cache
+   * an optional list of dependencies of the program on the input tensors. If this is not specified, the program depends
+   * on 'dims' of all inputs.
    */
-  cacheHint?: string;
+  inputDependencies?: ProgramInputTensorInfoDependency[];
 }
 
 /**
- * A ProgramInfoLoader allows
+ * Represent information about a program's cache for uniform.
  */
-export interface ProgramInfoLoader extends ProgramMetadata {
+export interface ProgramUniformCacheInfo {
+  /**
+   * an optional string as a cache hint in the uniform cache. If this is not specified, the cache hint will be empty.
+   *
+   * This hint string should only contains runtime information, such as the shape of inputs.
+   */
+  hint?: string;
+
   /**
-   * a function to get the program info
+   * an optional list of dependencies of the program on the input tensors. If this is not specified, the program depends
+   * on 'none' of all inputs.
    */
-  get(): ProgramInfo;
+  inputDependencies?: ProgramInputTensorInfoDependency[];
 }
 
+
 /**
  * A set of data that represent a shader program
  */
-export interface ProgramInfo extends ProgramMetadata {
+export interface ProgramInfo {
   /**
-   * information of uniform variables
+   * the name of the program. used for debugging and profiling
    */
-  variables?: ProgramVariable[];
+  name: string;
+
   /**
-   * tensor info for outputs
+   * gpu data types for each input
    */
-  outputs: TensorInfo[];
+  inputTypes: GpuDataType[];
```

Comment:
It seems that the `inputTypes` is not very useful. I only find it's used in https://github.com/microsoft/onnxruntime/blob/main/js/web/lib/wasm/jsep/backend-webgpu.ts#L224. Not sure whether we can remove it.